snuba/datasets/entities/clickhouse_upgrade.py

```diff
@@ -94,3 +117,179 @@ def choose(self, referrer: str) -> Choice:
+    - Calculate the average of the similarity scores of all the matching rows.
+    - Rows that do not have a match have a similarity score = 0.0
+    """
+    if primary_result is None or len(results) == 0:
+        return
+
+    secondary_result = results[0]
+    primary_schema = split_metadata(primary_result.result)
+    secondary_schema = split_metadata(secondary_result.result)
+
+    if primary_schema != secondary_schema:
+        metrics.increment("schema_mismatch", tags={"referrer": referrer})
+        return
+
+    if not primary_schema.complete or secondary_schema.complete:
+        metrics.increment("incomplete_match", tags={"referrer": referrer})
+
+    score = calculate_score(
+        primary_result.result.result["data"],
+        secondary_result.result.result["data"],
+        primary_schema,
+    )
+    metrics.timing("similarity_score", score, tags={"referrer": referrer})
+
+
+def calculate_score(
+    data_row1: MutableSequence[Row],
+    data_row2: MutableSequence[Row],
+    schema: SplitSchema,
+) -> float:
+    """
+    Calculate the similarity score of two lists of rows.
+    The total score is the average of the scores of each pair of mathcing
+    rows where rows are matching is they have the same key.
+    Rows that do not have a match count as 0.
+    The similarity score between two matching rows is the average of the
+    similarities between all columns.
+    """
+
+    # Each key have a list of rows associated because a given key is
+    # not necessarily unique within a result.
+    result1_index: MutableMapping[Key, List[SplitRow]] = defaultdict(list)
+    for i in range(min(len(data_row1), settings.MAX_ROWS_TO_CHECK_FOR_SIMILARITY)):
+        row = split_row(data_row1[i], schema)
+        result1_index[row.key].append(row)
+
+    row_similarities: MutableSequence[float] = []
+    for i in range(min(len(data_row2), settings.MAX_ROWS_TO_CHECK_FOR_SIMILARITY)):
+        row = split_row(data_row2[i], schema)
+        if row.key not in result1_index:
+            row_similarities.append(0.0)
+        else:
+            matching_rows = result1_index[row.key]
+            # When we find multiple rows in result1 that match a row in result2
+            # We pick the one with the highest score.
+            # We choose the highest match as that is most likely the row in
+            # result1 that correspond to the row we are examining in reult2.
+            key_similarities = [
+                row_similarity(row, matching_row) for matching_row in matching_rows
+            ]
+            row_similarities.append(max(key_similarities))
+
+    result2_index: MutableMapping[Key, List[SplitRow]] = defaultdict(list)
+    for i in range(min(len(data_row2), settings.MAX_ROWS_TO_CHECK_FOR_SIMILARITY)):
+        row = split_row(data_row2[i], schema)
+        result2_index[row.key].append(row)
+
+    for i in range(min(len(data_row1), settings.MAX_ROWS_TO_CHECK_FOR_SIMILARITY)):
+        row = split_row(data_row1[i], schema)
```

Comment:
This key has already been calculated and is available in `result1_index`. You could iterate through that instead.